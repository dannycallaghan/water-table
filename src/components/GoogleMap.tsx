import {
  APIProvider,
  Map,
  AdvancedMarker,
  useAdvancedMarkerRef,
  InfoWindow,
} from '@vis.gl/react-google-maps';
import formatcoords from 'formatcoords';
import { memo } from 'react';
import { useCallback, useState } from 'react';

function GoogleMap(props: {
  pins: { lat: string; lng: string; sensorId: string }[];
  zoom: number;
  height: string;
  width: string;
}) {
  let position = {
    lat: Number(props.pins[0]?.lat),
    lng: Number(props.pins[0]?.lng),
  };

  if (props.pins.length > 1) {
    position = {
      lat: 51.5072,
      lng: 0.1276,
    };
  }

  return (
    <>
      <APIProvider apiKey={import.meta.env.VITE_GOOGLE_API_KEY}>
        <Map
          mapId={'de2fd94a6060890b'}
          defaultCenter={position}
          style={{
            width: props.width,
            height: props.height,
          }}
          defaultZoom={props.zoom}
        >
          {props.pins.map((pin, i) => (
            <PinInfoWindow key={`${pin.lat}${pin.lng}${i}`} pin={pin} />
          ))}
        </Map>
      </APIProvider>
    </>
  );
}

const PinInfoWindow = memo(function PinInfoWindow(props: {
  pin: {
    lat: string;
    lng: string;
    sensorId: string;
  };
}) {
  const { pin } = props;
  const [markerRef, marker] = useAdvancedMarkerRef();
  const [infoWindowShown, setInfoWindowShown] = useState(false);

  const handleMarkerClick = useCallback(
    () => setInfoWindowShown((isShown) => !isShown),
    []
  );

  const handleClose = useCallback(() => setInfoWindowShown(false), []);

  return (
    <>
      <AdvancedMarker
        ref={markerRef}
        position={{ lat: Number(pin.lat), lng: Number(pin.lng) }}
        onClick={handleMarkerClick}
      />
      {infoWindowShown && (
        <InfoWindow anchor={marker} className="w-[300px]" onClose={handleClose}>
          <p className="text-sm text-gray-500">Identifer:</p>
          <p className="text-sm text-gray-900">{pin.sensorId}</p>
          <p className="text-sm text-gray-500">Location:</p>
          <p className="text-sm text-gray-900">
            {formatcoords(Number(pin.lat), Number(pin.lng)).format()}
          </p>
        </InfoWindow>
      )}
    </>
  );
});

export default GoogleMap;
