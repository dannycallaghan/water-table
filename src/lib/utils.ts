import { IMessage, IMessages } from '../types/types';
import { LOCALE } from './globals';

export function compareISODates(str1: string, str2: string): number {
  return new Date(str1).valueOf() - new Date(str2).valueOf();
}

export function isBetweenISODates(target: string, range: string[]): boolean {
  return compareISODates(target, range[0]) > 0
    && compareISODates(target, range[1]) < 0;
}

export function filterByDate(data: IMessages, range: string[]): IMessages {
  if (!Array.isArray(range)) return data;
  if (range.length < 2) return data;
  return data.filter(message => isBetweenISODates(message.transmittedAt.iso, range))
}

export function formatDateTime(datetime: string) {
  const stringAsDate: Date = new Date(datetime);
  const options: Intl.DateTimeFormatOptions = {
    year: 'numeric',
    month: 'short',
    day: 'numeric',
  };
  const date = stringAsDate.toLocaleDateString(LOCALE, options);
  const time = stringAsDate.toLocaleTimeString(LOCALE);
  return `${date} ${time}`;
}

export function b64DecodeUnicode(str: string) {
  return decodeURIComponent(
    Array.prototype.map
      .call(
        atob(str),
        (c) => '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2)
      )
      .join('')
  );
}

export function parseJWT(token: string) {
  return JSON.parse(b64DecodeUnicode(token));
}

export function getDataCoords(data: IMessages) {
  return data.map((item: IMessage) => ({
    lat: item.latitude,
    lng: item.longitude,
    sensorId: item.sensorId,
  }));
}
