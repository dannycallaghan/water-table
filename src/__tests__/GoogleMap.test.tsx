import sensorData from '../data/river-sensor-data-test.json';
import { expect, test, describe } from 'vitest';
import { render } from '@testing-library/react';
import { getDataCoords } from '../lib/utils.ts';
import GoogleMap from '../components/GoogleMap.tsx';

describe('GoogleMap', () => {
  test('renders correctly', () => {
    const { asFragment } = render(
      <GoogleMap
        pins={getDataCoords(sensorData.slice(0, 3))}
        height="225px"
        width="100%"
        zoom={5}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
