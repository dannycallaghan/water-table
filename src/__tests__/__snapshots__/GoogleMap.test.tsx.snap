// Vitest Snapshot v1

exports[`GoogleMap > renders correctly 1`] = `
<DocumentFragment>
  <div
    data-testid="map"
    style="width: 100%; height: 225px; position: relative; z-index: 0;"
  />
</DocumentFragment>
`;
